subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java

```diff
@@ -73,34 +77,62 @@ public RecompilationSpec provideRecompilationSpec(CurrentCompilation current, Pr
+        return CompileTransaction.newTransaction(tempDir, fileOperations, deleter)
+            .onFailureRollbackStashIfException(t -> t instanceof CompilationFailedException)
+            // Move files that should be deleted to new folders that will be reverted in case of a failure
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(classesToDelete, compileDestinationDir))
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(classesToDelete, annProcessorGeneratedSourcesDir))
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(classesToDelete, headerOutputDir))
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(resourcesToDelete.get(CLASS_OUTPUT), compileDestinationDir))
+            // If the client has not set a location for SOURCE_OUTPUT, javac outputs those files to the CLASS_OUTPUT directory, so delete that instead.
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(resourcesToDelete.get(SOURCE_OUTPUT), MoreObjects.firstNonNull(annProcessorGeneratedSourcesDir, compileDestinationDir)))
+            // In the same situation with NATIVE_HEADER_OUTPUT, javac just NPEs.  Don't bother.
+            .newTransactionalDirectory(dir -> dir.stashFilesForRollbackOnFailure(resourcesToDelete.get(GeneratedResource.Location.NATIVE_HEADER_OUTPUT), headerOutputDir))
+            // Set compile outputs to new folders and in case of a success move content to original folders
+            .newTransactionalDirectory(dir -> dir
+                .onSuccessMoveFilesTo(compileDestinationDir)
+                .ensureEmptyKeepingDirectoryStructureFrom(compileDestinationDir)
+                .beforeExecutionDo(() -> spec.setDestinationDir(dir.getAsFile()))
+                .afterExecutionAlwaysDo(() -> spec.setDestinationDir(compileDestinationDir))
+            )
+            .newTransactionalDirectory(annProcessorGeneratedSourcesDir != null, dir -> dir
+                .onSuccessMoveFilesTo(annProcessorGeneratedSourcesDir)
+                .ensureEmptyKeepingDirectoryStructureFrom(annProcessorGeneratedSourcesDir)
+                .beforeExecutionDo(() -> spec.getCompileOptions().setAnnotationProcessorGeneratedSourcesDirectory(dir.getAsFile()))
+                .afterExecutionAlwaysDo(() -> spec.getCompileOptions().setAnnotationProcessorGeneratedSourcesDirectory(annProcessorGeneratedSourcesDir))
+            )
+            .newTransactionalDirectory(headerOutputDir != null, dir -> dir
+                .onSuccessMoveFilesTo(headerOutputDir)
+                .ensureEmptyKeepingDirectoryStructureFrom(headerOutputDir)
+                .beforeExecutionDo(() -> spec.getCompileOptions().setHeaderOutputDirectory(dir.getAsFile()))
+                .afterExecutionAlwaysDo(() -> spec.getCompileOptions().setHeaderOutputDirectory(headerOutputDir))
+            );
```

Comment:
It seems that we are doing two things here: we are either stashing some files to be restored on failure, or we are we are preparing an empty directory to be copied back on success. I think we could have two methods for these two use cases on `CompileTransaction`. I don't see what the builder approach adds here.

Reply:
Simplified that. I kept the builder since it make sense to me, but we have just two methods on builder and it's really a builder (so it outputs immutable object):
```
CompileTransactionBuilder stashFiles(PatternSet filesPattern, File sourceDirectory) {
CompileTransactionBuilder stageOutputDirectory(File sourceDirectory, Consumer<File> setSpecOutputFunction) {
```

I think it makes sense to keep builder since you have to:
1. list folder with patterns for files you want to stash
2. you also have to list output folders that you want to stage.

I could move some logic out of `CompileTransaction`, to maybe have just method like:
```
CompileTransaction.execute(
    List<File> filesToStash,
    // pair of "staging" output dir and original output dir
    List<Pair<File, File>> outputFolders, 
    Action<> action
)
```
But it makes sense to me, that transaction handles all about finding files, cleaning folders and not just restoring. Also once you try to test it, it makes sense to test all this together.

Reply:
After some more thoughts I went with the _alternative_ solution from comment above and totally remove builder.